### DISCUSSION TITLE
browser integration with sandboxing enabled

### DISCUSSION THREAD

**Discussion by D0nnieD4rk0** [keepassxreboot/keepassxc#6741] [Q&A] [answered]
recently started playing around with firejail and im wondering what files/binaries are required to be whitelisted for the browser extension to see the client

**Comment by droidmonkey** [comment, marked as answer]
You need to allow the browser to start the proxy application and connect to the local socket. Socket location can be found at https://github.com/keepassxreboot/keepassxc/blob/089c8df01d1676abb6d084edf4da7038bc0d9be2/src/browser/BrowserShared.cpp#L34

Your best bet is to strace your browser and look for failures.

**Comment by rusty-snake** [comment]
All you need to do: https://github.com/netblue30/firejail/blob/0.9.66/etc/profile-a-l/firefox.profile#L27-L30

See https://github.com/netblue30/firejail/issues/3952 and https://github.com/netblue30/firejail/issues/4279 for more information.

**Reply by rusty-snake** [reply]
Since we are already talking about keepassxc-browser+sandbox, it would be awesome if KPXC would use `$XDG_RUNTIME_DIR/keepassxc/socket` (i.e. a subdir of `$XDG_RUNTIME_DIR`) instead of `$XDG_RUNTIME_DIR/socket`.
If there is a chance that this gets implemented I can open an issue for it.

**Reply by droidmonkey** [reply]
For what reason? We already use our reverse domain naming convention. Although I think we need to implement that convention for the single instance socket.

**Reply by rusty-snake** [reply]
#### Background

It starts with the fact that you don't want to start firefox and keepassxc in the same sandbox. Firefox is likely the biggest (remote) attack vector on your system (after you :wink:) and KeePassXC keeps (all) you secrets. Therefore you have two sandboxes (one for ff and one for kpxc).

As the goal of sandboxing is to have as much isolation between programs as possible, you start with blacklisting. You deny any access to files known to contain secrets / be insecure or which the sandboxed program just don't need. However this is not a perfect solution as you never know all files out there.

Therefore you do whitelisting. You allow the sandboxed program only to access (and see) files which are known to be needed/used.

<sub>*While you should use better names for (new) implementations [1](https://en.wikipedia.org/wiki/Blacklist_(computing)#Controversy_over_use_of_the_term), [2](https://inclusivenaming.org/), [3](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/process/coding-style.rst#n349), there is still a lack of widely used and uniform names to talk about the concept of blacklisting/whitelisting IMHO.*</sub>

#### Whitelisting

The implementation of whitelisting depends (firejail, systemd, bubblewrap based, ...) but in the end it works always by mounting a private filesystem (e.g. a `tmpfs`) at `$XDG_RUNTIME_DIR` and bind-mounting the necessary paths.

This results in a few limitations:
- You can not rename/move a whitelisted path. (What you can is to rename/move inside a whitelisted path).
- In order to whitelist a path, it must exist when you create the sandbox.
- If the whitelisted path is a socket and it is deleted and recreated outside the sandbox, the sandbox will still have the old (now unconnected) socket.

#### The problem with keepassxc-browser and whitelisting

Because the socket that kpxc creates needs to accessible by firefox, you are either restricted to use only blacklisting for both sandboxes or to use whitelisting for the firefox sandbox and start kpxc before firefox and keep it open.

If the keepassxc-browser socket would be in a subdir of `$XDG_RUNTIME_DIR`, this directory could be created before the sandbox is started (it's just a directory) and whitelisted. Then firefox could use sockets created by a keepassxc instance that is start after firefox.

**TL;DR:** This would allows tight sandbox without restricting the workflow of the user (start/stop ff and kpxc when you want in any order).

**Reply by droidmonkey** [reply]
Couldn't you just `noblacklist` the keepassxc socket files and `blacklist` the {RUNUSER} directory? Then everything would be blocked except the keepassxc files without explicitly whitelisting them (which I gather requires them to exist). I am not a firejail expert, just reading the docs. 

There is also the option to mkfile similar to mkdir, not sure if that helps either. 

There must be a way to handle files that are not present at startup.

**Reply by rusty-snake** [reply]
> Couldn't you just noblacklist the keepassxc socket files and blacklist the {RUNUSER} directory? 

The following does not work:

```
noblacklist ${RUNUSER}/org.keepassxc.KeePassXC.BrowserServer
blacklist ${RUNUSER}
```

What would work:

```
noblacklist ${RUNUSER}/bus
noblacklist ${RUNUSER}/dconf
noblacklist ${RUNUSER}/gdm/Xauthority
noblacklist ${RUNUSER}/ICEauthority
noblacklist ${RUNUSER}/.mutter-Xwaylandauth.*
noblacklist ${RUNUSER}/pulse/native
noblacklist ${RUNUSER}/wayland-0
noblacklist ${RUNUSER}/wayland-1
noblacklist ${RUNUSER}/xauth_*
noblacklist ${RUNUSER}/[[:xdigit:]][[:xdigit:]][[:xdigit:]][[:xdigit:]][[:xdigit:]][[:xdigit:]][[:xdigit:]][[:xdigit:]]-[[:xdigit:]][[:xdigit:]][[:xdigit:]][[:xdigit:]]-[[:xdigit:]][[:xdigit:]][[:xdigit:]][[:xdigit:]]-[[:xdigit:]][[:xdigit:]][[:xdigit:]][[:xdigit:]]-[[:xdigit:]][[:xdigit:]][[:xdigit:]][[:xdigit:]][[:xdigit:]][[:xdigit:]][[:xdigit:]][[:xdigit:]][[:xdigit:]][[:xdigit:]][[:xdigit:]][[:xdigit:]]
noblacklist ${RUNUSER}/org.keepassxc.KeePassXC.BrowserServer
blacklist ${RUNUSER}/*
```

However this is still less secure than whitelisting because *anything* created after the startup will still appear.

> There is also the option to mkfile similar to mkdir, not sure if that helps either.

For regular files which are `open`/`read`+`write`/`close` `mkfile` is fine but for a sockets I don't think that it works (~though I am not 100% sure~ edit: tested, `mkfile` does not work with sockets).

> There must be a way to handle files that are not present at startup.

If you mean that you belief there is a way: no, at least for firejail everything is done at startup.

If you mean that this should be implemented:
- For `blacklist` this is racey.
- For `whitelist` this could be done using `inotify`. However with the current development speed (for new feature) this will take years, therefore I asked here if there is a chance this will be done. *(From a external point of view this looks like a small change to KeePassXC (compared to a bigger change in firejail)).* Furthermore a change in firejail would not help users of other sandboxing programs (whether they are recommended or not is a separate question).

**Reply by droidmonkey** [reply]
It is a small change for us but I wanted to understand the limitations and semantics of firejail for future reference.

**Reply by rusty-snake** [reply]
FWIW I've hacked this my way.

<details><summary>firefox.local</summary>

```
mkdir ${RUNUSER}/keepassxc
whitelist ${RUNUSER}/keepassxc
# Only if private-bin is enabled for firefox
private-bin keepassxc-proxy,keepassxc-proxy-wrapper
```

</details>
<details><summary>keepassxc.local</summary>

```
# Only if whitelisting is enabled for keepassxc
mkdir ${HOME}/.mozilla/native-messaging-hosts
mkfile ${HOME}/.mozilla/native-messaging-hosts/org.keepassxc.keepassxc_browser.json
whitelist ${HOME}/.mozilla/native-messaging-hosts/org.keepassxc.keepassxc_browser.json
# Does not work with GNOME XWayland.
# In this form only for keepassxc --platform wayland and UID 1000.
mkdir /run/user/1000/keepassxc
whitelist /run/user/1000/keepassxc
env XDG_RUNTIME_DIR=/run/user/1000/keepassxc
env WAYLAND_DISPLAY=/run/user/1000/wayland-0
#env DBUS_SESSION_BUS_ADDRESS is set by dbus-user filter
```

</details>
<details><summary>keepassxc.ini</summary>

```ini
[Browser]
CustomProxyLocation=/usr/local/bin/keepassxc-proxy-wrapper
Enabled=true
UseCustomProxy=true
```

</details>
<details><summary>/usr/local/bin/keepassxc-proxy-wrapper</summary>

```bash
#!/bin/bash
export XDG_RUNTIME_DIR=/run/user/1000/keepassxc
exec /usr/bin/keepassxc-proxy
```

</details>

**Reply by droidmonkey** [reply]
That seems legit!
